Change the way the TickerTask tracks Blocks

md5sha256 opened the pull request:
## Description
<!-- Please explain why you are making this pull request. -->
<!-- Start writing below this line -->
As per a discussion in the SF discord, this PR removes the use of explicit locks used in TickerTask. 
I'm not a massive fan of the implementation of `TickingBlocks` as it is right now. Currently its just been wrapped (mutex) so it isn't that much better for performance than it was before (might even be worse on that end). However, we now map the world to a long as opposed to `ChunkPosition --> Set<Location>` which means the memory footprint is considerably lower. 

## Proposed changes
<!-- Please explain what changes you have made to the code. -->
<!-- Start writing below this line -->
- Add `TickingBlocks`, a class that serves as an accessor/wrapper for a `Map<World, Set<Long>>`. In other words, a `BlockPositionSet`. 
- Removes the public getters for the `tickingLocations`
- Swap out the `Lock` + `Map<ChunkPosition, Set<Location>>` for `TickingBlocks`
- Added relevant docs

## Related Issues (if applicable)
<!-- Please tag any Issues related to your Pull Request -->
<!-- Syntax: "Resolves #000" -->
<!-- Start writing below this line -->

## Checklist
<!-- Here is a little checklist you can follow. -->
<!-- Click on these checkboxes after you created the pull request. -->
<!-- Don't worry, these are not requirements. They only serve as guidance. -->
- [ ] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

TheBusyBiscuit commented on src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickingBlocks.java:
```diff
@@ -0,0 +1,165 @@
+ *
+ * @author md5sha256
+ */
+final class TickingBlocks {
```
Not a fan of this name tbh.
Perhaps `TickerCollection`?

TheBusyBiscuit commented on src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickingBlocks.java:
```diff
@@ -0,0 +1,165 @@
+     * @see #getBlocksCopy()
+     */
+    public @Nonnull Set<Long> getBlocks(@Nonnull World world) {
+        return blocks.computeIfAbsent(world, (unused) -> Collections.synchronizedSet(new HashSet<>()));
```
Don't put braces around a single variable in a lambda expression
```suggestion
        return blocks.computeIfAbsent(world, x -> Collections.synchronizedSet(new HashSet<>()));
```

TheBusyBiscuit commented on src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickingBlocks.java:
```diff
@@ -0,0 +1,165 @@
+        return locations;
+    }
+
+    public void addBlock(@Nonnull BlockPosition block) {
+        getBlocks(block.getWorld()).add(block.getPosition());
+    }
+
+    public void addBlock(@Nonnull Location location) {
+        addBlock(new BlockPosition(location));
+    }
+
+    public void removeBlock(@Nonnull BlockPosition block) {
+        World world = block.getWorld();
+        Collection<Long> positions = blocks.get(world);
+        if (positions == null) {
+            return;
+        }
+        positions.remove(block.getPosition());
+        if (positions.isEmpty()) {
+            blocks.remove(world);
+        }
+    }
+
+    public void removeBlock(@Nonnull Location location) {
+        removeBlock(new BlockPosition(location));
+    }
```
sum javadocs? :D

(No necessity, just nice to have)

TheBusyBiscuit commented on src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickingBlocks.java:
```diff
@@ -0,0 +1,165 @@
+     * @return Returns a {@link Set} of {@link BlockPosition BlockPositions}
+     * @see #getBlocksCopy()
+     */
+    public @Nonnull Set<BlockPosition> getBlocksCopy(@Nonnull World world) {
+        Collection<Long> compressed = blocks.get(world);
+        if (compressed == null) {
+            return Collections.emptySet();
+        }
+        return convertToBlockPosition(world, compressed);
+    }
+
+    /**
+     * Get a copy of all the blocks stored in the form of {@link BlockPosition BlockPositions}
+     * The returned {@link Set} is not guaranteed to be thread-safe.
+     *
+     * @return Returns a {@link Set} of {@link BlockPosition BlockPositions}
+     * @see #getBlocksCopy(World)
+     */
+    public @Nonnull Set<BlockPosition> getBlocksCopy() {
+        Set<BlockPosition> positions = new HashSet<>();
+        for (Map.Entry<World, Set<Long>> entry : blocks.entrySet()) {
+            World world = entry.getKey();
+            Set<BlockPosition> converted = convertToBlockPosition(world, entry.getValue());
+            positions.addAll(converted);
+        }
+        return positions;
+    }
```
These methods seem to be unused and are public, yet this class isn't even public and the TickerTask provides no getter for this object anyway.

md5sha256 commented:
I've resolved the comments. That being said I've also changed the implementation of `TickingCollection` to use a guava `Cache` as that offers better concurrent performance whilst also supporting weak keys. Alongside biscuit, could @WalshyDev also have a look through?

TheBusyBiscuit commented:
> Alongside biscuit, could @WalshyDev also have a look through?

I am not enough :FeelsBadMan:

md5sha256 commented:
> > Alongside biscuit, could @WalshyDev also have a look through?
> 
> I am not enough :FeelsBadMan:

Sorry :P Just thought Walshy's experience with the guava Cache might make a difference since we've discussed how to implement this for a bit back in April and early May

md5sha256 commented:
👀

WalshyDev commented on src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:
```diff
@@ -0,0 +1,129 @@
+ *
+ * @author md5sha256
+ */
+final class TickerCollection {
```
We should annotate with @ThreadSafe

WalshyDev reviewed (commented):
Looking good. Has this been stress tested?

(also apologies, didn't realise this was waiting on me)

md5sha256 commented:
No, it has not. I can start working on a JMH bench. I just needed an opinion on the use of the Guava Cache as I feel like I'm abusing it here.

WalshyDev commented:
Hmmm,Guava's Cache is made and optimised for automatic eviction (usually size or time). We don't really use any of the Cache's features at all nor use it for the intended purpose. I feel like we're better off just using ConcurrentHashMap.

The only feature/benefit we get from Cache is the weakKeys. Considering the compute and retrieval of things isn't expensive and memory is more valuable to MC servers than usual production software I think we stick to native Java ConcurrentHashMap.

md5sha256 commented:
Noted. So should I wrap all keys with a weak reference?

WalshyDev commented:
Probably best yeah

md5sha256 commented:
I guess I have to stick with holding the UUID instance of the world instead?

WalshyDev commented:
Yeah UUID would be better.

md5sha256 commented:
Also, by stress test do you mean a unit test that stresses the collection to check for concurrency errors or a benchmark?

WalshyDev commented:
> Also, by stress test do you mean a unit test that stresses the collection to check for concurrency errors or a benchmark?

Making sure it can handle concurrency correctly.

WalshyDev commented on src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:
```diff
@@ -0,0 +1,134 @@
+package io.github.thebusybiscuit.slimefun4.implementation.tasks;
+
+
```
remove double new line

WalshyDev commented on src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java:
```diff
@@ -75,10 +66,11 @@
      */
     private static final long MAX_WAIT_TIME = 1_500_000L;
 
+
```
double new line

md5sha256 commented:
Should be good EXCEPT for https://discord.com/channels/565557184348422174/573872826327629844/856183240736899104. 
The tests currently pass as I've modified the test to not fail (I changed it to assert against a fresh `Set` from `#getBlocks` instead of `compressedPositions`. The stress test passes; however, I'm all eyes and ears for whether the stress test needs more work. 
Tested on an Intel i7-7700 (4 cores, 8 threads).

md5sha256 commented:
Also, I was forced to merge in `master` because I couldn't compile the new spelling tests as the localisation stuff changed. Oh well.

WalshyDev commented on src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    /**
+     * Remove a location from this ticker collection.
```
```suggestion
     * Remove a {@link Location} from this ticker collection.
```

WalshyDev commented on src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:
```diff
@@ -0,0 +1,133 @@
+    /**
+     * Remove a location from this ticker collection.
+     *
+     * @param location The location of the block to remove
```
```suggestion
     * @param location The {@link Location} of the block to remove
```

WalshyDev commented on src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    /**
+     * Clear all the blocks for a given {@link World}
+     * @param key The world to clear the key for
```
```suggestion
     * Clear all the blocks for a given {@link World}
     *
     * @param key The {@link World} to remove blocks from
```

WalshyDev commented on src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java:
```diff
@@ -101,17 +92,19 @@
+
     /**
-     * Making collections concurrent doesn't make their iterators thread safe.
+     *  These are "Queues" of blocks that need to be removed or moved.
+     *  <p>
+     *  Whilst the {@link ConcurrentHashMap} implementation is thread-safe,
+     *  the {@link Location Locations} are not. Therefore, the internal copy of the locations
+     *  should be treated as <strong>read only</strong> and never exposed externally.
```
2 spaces at the start here
```suggestion
     * These are "Queues" of blocks that need to be removed or moved.
     * <p>
     * Whilst the {@link ConcurrentHashMap} implementation is thread-safe,
     * the {@link Location Locations} are not. Therefore, the internal copy of the locations
     * should be treated as <strong>read only</strong> and never exposed externally.
```

WalshyDev commented on src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java:
```diff
@@ -338,22 +320,21 @@ private void processAsyncTasks() {
-                        reportErrors(l, item, x);
-                    }
+    /**
+     * Executes all tasks in a given {@link Collection} of locations
```
```suggestion
     * Executes all tasks in a given {@link Collection} of {@link Location locations}
```

md5sha256 commented on src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:
```diff
@@ -0,0 +1,133 @@
+     */
+    public @Nonnull Set<Location> getBlocksAsLocations() {
+        Set<Location> locations = new HashSet<>();
+        blocks.keySet().removeIf(uid -> Bukkit.getWorld(uid) == null);
+        for (Map.Entry<UUID, Set<Long>> entry : blocks.entrySet()) {
+            World world = Bukkit.getWorld(entry.getKey());
```
Won't that throw a concurrent modification exception, when removing keys whilst iterating over the entryset?

WalshyDev replied:
Nope, it's thread-safe.

md5sha256 replied:
I guess, but won't that also cause the original iterator the for-loop is using to have undertiministic behaviour?

WalshyDev replied:
Hmmm yeah fair enough.

md5sha256 commented on src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:
```diff
@@ -0,0 +1,133 @@
+
+    private static @Nonnull Set<Location> convertToLocation(@Nonnull World world, @Nonnull Collection<Long> raw) {
+        Set<Location> locations = new HashSet<>(raw.size());
+        for (long compressed : raw) {
+            int x = (int) (compressed >> 38);
+            int y = (int) (compressed & 0XFFF);
+            int z = (int) (compressed << 26 >> 38);
+            locations.add(new Location(world, x, y, z));
+        }
```
I'll PR that to cscl2 and wait for a bump.

md5sha256 replied:
Righty then, now waiting on https://github.com/TheBusyBiscuit/CS-CoreLib2/pull/194

WalshyDev reviewed (approved):
LGTM

TheBusyBiscuit reviewed (approved):
What Walshy said 🤷🏻
